stackrox/stackrox
# ROX-20183: Add conditional RBAC for violation tabs

## Description

Follow up #7411

### Problem

1. Conditional rendering for **Deployment** tab to prevent request error without Deployment resource was unable to display 5 properties from `deployment` property of alert response pending component refactoring.
2. **Deployment tab** renders **Network policies** card:
    * That request requires NetworkPolicy instead of Deployment resource, therefore might be omitted needlessly when the tab is omitted.
    * Network policies are for a **namespace**, not a particular deployment.

### Analysis

1. ViolationDetailsPage.tsx file:
    * Requires Deployment resource to render **Deployment** tab.
2. Deployment/DeploymentDetails.tsx file:
    * `NetworkPoliciesCard` element uses `alertDeployment` object.
    * Other elements use `relatedDeployment` object from `useFetchDeployment` hook.
3. Deployment/DeploymentOverview.tsx file:
    * Could display the first 5 items with `alertDeployment` object properties.
    * Display the others with `relatedDeployment` object properties.
        **Labels** and **Annotations** are in a gray area. In theory, could display with `alertDeployment` but one example I found had inconsistent values of `labels` between the requests.

### Solution

1. Update integration tests for **Network policies** tab.
2. Edit ViolationDetailsPage.tsx file:
    * Change conditional rendering for **Deployment** tab to alternative elements with and without read access for Deployment resource.
    * Lift up **Network policies** card from **Deployment** tab to its own tab.
3. Split Deployment/DeploymentDetails.tsx file:
    * Deployment/DeploymentTabWithReadAccessForDeployment.tsx
    * Deployment/DeploymentTabWithoutReadAccessForDeployment.tsx
4. Edit DeploymentOverview.tsx file:
    * Add required `alertDeployment: TODO` prop and make `deployment?: Deployment` prop optional.
    * Replace `deployment` with `alertDeployment` in first 5 items.
    * Conditionally render other items.
5. Move and rename NetworkPoliciesCard.tsx file:
    * NetworkPolicies/NetworkPoliciesTab.tsx
6. Move NetworkPolicyModal.tsx to NetworkPolicies folder.
7. Add `AlertDeployment` and `AlertResource` types in frontend storage types.

## Checklist

## Testing Performed

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.js: 0 = 3910659 - 3910659
        total: 924 = 10162367 - 10161443
    * `ls -al build/static/js/*.js | wc -l`
        0 = 83 - 83 files
3. `yarn start` in ui

### Manual testing

Pictures at 1440px width and 900px height like a newer laptop.

1. Visit /main/violations and click a row link to visit a violation page.

2. Click **Deployment** tab.

    * With Deployment resource and scroll down to see absence of **Network policies** card.
        ![DeploymentTabWithReadAccessForDeployment](https://github.com/stackrox/stackrox/assets/11862657/c2972619-fa7b-4d75-9823-d2feee96d243)

    * Without Deployment resource.
        ![DeploymentTabWithoutReadAccessForDeployment](https://github.com/stackrox/stackrox/assets/11862657/b13e929f-fcc0-4518-b201-863e1951675a)

3. Click **Network policies** tab.
    ![NetworkPoliciesTab](https://github.com/stackrox/stackrox/assets/11862657/2b516217-683c-40df-978f-6dbd32959b66)

4. Click namespace link to open modal.
    ![NetworkPolicyModal](https://github.com/stackrox/stackrox/assets/11862657/adcccf5a-f4c5-4d09-850f-6b7bee25b8e4)

5. See absence of **Network policies** tab without NetworkPolicy resource.

### Integration testing

* violations/violations.test.js
